### PR TITLE
#23 poc translator

### PR DESCRIPTION
**Pull request by qubitter**

### CODE REVIEW

File: oxy/poc_translator.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    result.insert(4, data[6] as f32);
+    result.insert(5, data[7] as f32);
+    result
+}
+'''
+out_string: str = ""
+data = yaml.load(open("mapping.yaml"))[1]
+# print(data)
+
+function_name: str = "decode" + "_" + "_".join(data['string'].split(" "))
```
Comment (Peyton-McKee) [inline oxy/poc_translator.py:25]: why not just “decode_” + “_”.join(...

File: oxy/poc_translator.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+signature: str = "pub fn " + function_name + args + returnVal + " {"
+instantiate_hash_map: str = "    let mut result = HashMap::new();"
+conclusion: str = "    result\n}"
+
+decodings: list[str] = []
+accumulated_size: int = 0
+for field in data["fields"]: # result.insert(1, (pd::big_endian(&data[0..2], 8) as f32) / 10.0);
+    field: dict = field
+    id = field["field_id"]
+    if field["size"] > 1: # we need to do some decoding, then
```
Comment (Peyton-McKee) [inline oxy/poc_translator.py:38]: size = field[“size”]

File: oxy/poc_translator.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        to_decode: str = f"&data[{accumulated_size}..{accumulated_size+field['size']}]"
+        _cf: str = field.get("correcting_factor", "")
```
Comment (Peyton-McKee) [inline oxy/poc_translator.py:40]: correcting_factor_value =

File: oxy/poc_translator.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        correcting_factor: str = f"{(' / ' if '/' in _cf else ' * ') if _cf else ''}{_cf.split('/')[-1]}"
```
Comment (Peyton-McKee) [inline oxy/poc_translator.py:41]: oh yikes. yeah ok we need to do correcting_factor_op and correcting_factor_value in the yaml. Not this

File: oxy/poc_translator.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+'''
+out_string: str = ""
+data = yaml.load(open("mapping.yaml"))[1]
+# print(data)
+
+function_name: str = "decode" + "_" + "_".join(data['string'].split(" "))
+args: str = "(data: &[u8])"
+returnVal: str= " -> HashMap<u8, f32>"
+
+signature: str = "pub fn " + function_name + args + returnVal + " {"
```
Comment (Peyton-McKee) [inline oxy/poc_translator.py:29]: this is so funny.

File: oxy/poc_translator.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        to_decode: str = f"&data[{accumulated_size}..{accumulated_size+field['size']}]"
+        _cf: str = field.get("correcting_factor", "")
+        correcting_factor: str = f"{(' / ' if '/' in _cf else ' * ') if _cf else ''}{_cf.split('/')[-1]}"
+        for decodingsetup in field["decoding"]:
+            decodingsetup: dict[str, dict[str, str]] = {k: reduce(lambda x,y: x|y, v, {}) for k,v in decodingsetup.items()}
+            for decoder, params in decodingsetup.items():
+                match decoder:
+                    case "big_endian":
+                        to_decode = f"pd::big_endian({to_decode}, {params['bits']}) as {params['final_type']}"
+                    case "twos_complement":
```
Comment (Peyton-McKee) [inline oxy/poc_translator.py:48]: lots more cases but yes

File: oxy/poc_translator.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+accumulated_size: int = 0
+for field in data["fields"]: # result.insert(1, (pd::big_endian(&data[0..2], 8) as f32) / 10.0);
+    field: dict = field
+    id = field["field_id"]
+    if field["size"] > 1: # we need to do some decoding, then
+        to_decode: str = f"&data[{accumulated_size}..{accumulated_size+field['size']}]"
+        _cf: str = field.get("correcting_factor", "")
+        correcting_factor: str = f"{(' / ' if '/' in _cf else ' * ') if _cf else ''}{_cf.split('/')[-1]}"
+        for decodingsetup in field["decoding"]:
+            decodingsetup: dict[str, dict[str, str]] = {k: reduce(lambda x,y: x|y, v, {}) for k,v in decodingsetup.items()}
```
Comment (Peyton-McKee) [inline oxy/poc_translator.py:43]: sure but this is not readable so u either need to add a comment saying what this is doing or rename variables

File: oxy/poc_translator.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+            for decoder, params in decodingsetup.items():
+                match decoder:
+                    case "big_endian":
+                        to_decode = f"pd::big_endian({to_decode}, {params['bits']}) as {params['final_type']}"
```
Comment (Peyton-McKee) [inline oxy/poc_translator.py:47]: outside of the match assign bits = params[‘bits’] and final_type = params[‘final_type’]

File: oxy/poc_translator.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    id = field["field_id"]
+    if field["size"] > 1: # we need to do some decoding, then
+        to_decode: str = f"&data[{accumulated_size}..{accumulated_size+field['size']}]"
+        _cf: str = field.get("correcting_factor", "")
+        correcting_factor: str = f"{(' / ' if '/' in _cf else ' * ') if _cf else ''}{_cf.split('/')[-1]}"
+        for decodingsetup in field["decoding"]:
+            decodingsetup: dict[str, dict[str, str]] = {k: reduce(lambda x,y: x|y, v, {}) for k,v in decodingsetup.items()}
+            for decoder, params in decodingsetup.items():
+                match decoder:
+                    case "big_endian":
```
Comment (Peyton-McKee) [inline oxy/poc_translator.py:46]: also if ur gonna do this just u dont even need the match technically just do f”pd::{decoder}({to_decode}, {bits}) {final_type}

Comment (Peyton-McKee) [reply oxy/poc_translator.py:46]: just rename function names to match this in the rust code

Comment (Peyton-McKee) [reply oxy/poc_translator.py:46]: or change yaml to match function names

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]
impressive stuff. i find this really interesting code. Just have some clarifying comments and readibility comments

**Comment by nwdepatie** [comment]
Uh really quick it looks like your making a PR from your local main to remote main, typically we use remote branches so that other devs can easily checkout and verify code. Please make a remote branch and push to that to make a PR

File: oxy/mapping.yaml [hunk 10]
```diff
@@ -0,0 +1,43 @@
+1:
```
Comment (nwdepatie) [inline oxy/mapping.yaml:1]: I think we def need to think about how we structure this YAML. It functions as a source of truth of CAN messages on the car and should therefore be extremely human readable and parseable, such that any engineer can look through. I think as we expand, we need to consider both directory structure for within this repo as well as how the actual file is structured. How _most_ companies do this is that they have a ginormous doc of all messages to fully document the protocol (Amazon did this) or have a really big C file that everything includes that defines types (saw this at SimpliSafe, which is the way I originally tried to implement this stuff). TLDR this  is a very important topic to think about moving forward and should not be glossed over

File: oxy/poc_translator.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
+from ruamel.yaml import YAML
+from functools import reduce
+
+
+yaml: YAML = YAML(typ="safe")
+
+'''
+pub fn decode_accumulator_status(data: &[u8]) -> HashMap<u8, f32> {
```
[the remark this answers is not in the document]
Comment (nwdepatie) [reply oxy/poc_translator.py:8]: I'd assume it is a sample function that would be generated, but yeah definitely make that more explicit

File: oxy/poc_translator.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    result.insert(
+        2,
+        pd::twos_comp(pd::big_endian(&data[2..4], 8) as u32, 16) as f32 / 10.0,
+    );
+    result.insert(3, pd::big_endian(&data[4..6], 8) as f32);
+    result.insert(4, data[6] as f32);
+    result.insert(5, data[7] as f32);
+    result
+}
+'''
```
Comment (nwdepatie) [inline oxy/poc_translator.py:20]: Right off the bat this whole thing should be encapsulated in a class. I think there is a much better way of structuring this program around objects that would make it all easier to read and understand

File: oxy/poc_translator.py (hunk 9 above)
Comment (nwdepatie) [inline oxy/poc_translator.py:46]: This feels like too many indents, def think about splitting into functions to make this more readable. See above comment on classes to encapsulate logic.

File: oxy/poc_translator.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+out_string: str = ""
+data = yaml.load(open("mapping.yaml"))[1]
```
Comment (nwdepatie) [inline oxy/poc_translator.py:22]: So per the nature of YAML and kinda this general application, I think that you should really think about the data structures going into this. The problem feels a lot like it is parsing a bunch of very similar handlers that are all _slightly_ different, meaning you can probably have a class or something per attribute of the handler. Probably also want to create a "CAN message" class that contains all the attributes associated with a CAN message found in the YAML and then use that to pass to a general decoder.

File: oxy/poc_translator.py [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+# print(data)
+
+function_name: str = "decode" + "_" + "_".join(data['string'].split(" "))
+args: str = "(data: &[u8])"
+returnVal: str= " -> HashMap<u8, f32>"
+
+signature: str = "pub fn " + function_name + args + returnVal + " {"
+instantiate_hash_map: str = "    let mut result = HashMap::new();"
+conclusion: str = "    result\n}"
```
Comment (nwdepatie) [inline oxy/poc_translator.py:31]: Similar to the comment on using classes to represent the data in the YAML file, I think each component of the rust code would def benefit from putting it into a class.

**Review by nwdepatie — CHANGES_REQUESTED** [review]
Definitely a good start, but please consider scalability and readability in your design. I haven't tried running it so I am assuming you tested it locally and that it can generate code that compiles, but the way you are representing data is a bit lacking, and you didn't use any functions or classes to encapsulate functionality. This results in a lot of nested loops and global variables, in addition to not having a central entry point per regular python stuff. Definitely add in a main function via `if __name__ == "main"` or whatever I forget off the top of my head, and at least make a `YAMLParser` class and a `RustGenerator` class that passes around `CanMsg` objects. I think this is very much an object oriented problem and there is a lot of room to fuck around with some cool data types

**Comment by qubitter** [comment]
Closing to rephrase as a new branch, instead (yay, git)

**Pull request closed without merging**